cmd/utils.go

```diff
@@ -126,6 +126,92 @@ func parseEnvURL(envURL string) (*url.URL, string, string, *probe.Error) {
 	return u, accessKey, secretKey, nil
 }
 
+// probeS3Signature - auto probe S3 server signature: issue a Stat call
+// using v4 signature then v2 in case of failure.
+func probeS3Signature(accessKey, secretKey, url string) (string, *probe.Error) {
+	// Test s3 connection for API auto probe
+	s3Config := &Config{
+		// S3 connection parameters
+		Insecure:  globalInsecure,
+		AccessKey: accessKey,
+		SecretKey: secretKey,
+		Signature: "s3v4",
+		HostURL:   urlJoinPath(url, "probe-bucket-sign"),
+	}
+
+	s3Client, err := s3New(s3Config)
+	if err != nil {
+		return "", err
+	}
+
+	if _, err = s3Client.Stat(false, false); err != nil {
+		switch err.ToGoError().(type) {
+		case BucketDoesNotExist:
+			// Bucket doesn't exist, means signature probing worked V4.
+		default:
+			// Attempt with signature v2, since v4 seem to have failed.
+			s3Config.Signature = "s3v2"
+			s3Client, err = s3New(s3Config)
+			if err != nil {
+				return "", err
+			}
+			if _, err = s3Client.Stat(false, false); err != nil {
+				switch err.ToGoError().(type) {
+				case BucketDoesNotExist:
+					// Bucket doesn't exist, means signature probing worked with V2.
+				default:
+					return "", err
+				}
+			}
```

Comment:
In the event that "probe-bucket-sign" exists and server supports only s3v2, the api would erroneously use s3v4 in the config. Need to set to s3v2 if second Stat call doesn't return an error.

Reply:
why would that happen? s3Config.Signature is set to "s3v2"

Reply:
you are right. missed that.

Reply:
> In the event that "probe-bucket-sign" exists and server supports only s3v2, the api would erroneously use s3v4 in the config.

s3v2 signature is already set here: https://github.com/minio/mc/pull/2329/commits/2358b3bf6ae919e15409a2d5cd605f457fd9f7a8#diff-b4695c7b6902e27273329af6261b0c73R153